File: bikram-calendar/src/bikram.ts

/*
 * TypeScript implementation of Bikram calendar conversion
 * Copyright (c) 2024 onwards - khumnath cg, nath.khum@gmail.com
 * Licensed under the GNU General Public License v3.0 or later.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import { NP_MONTHS_DATA, BS_END_YEAR, BS_START_YEAR} from './data';
export class BikramDate {
  private year: number;
  private month: number; // 1-12
  private day: number;
  private readonly englishDate: Date;

  // Astronomical calculation constants (from the C++ implementation)
  private readonly YugaRotation_star = 1582237828;
  private readonly YugaRotation_sun = 4320000;
  private readonly YugaCivilDays: number;
  private readonly PlanetApogee_sun = 77 + 17.0 / 60;
  private readonly PlanetCircumm_sun = 13 + 50.0 / 60;
  private readonly rad = 57.2957795; // 180 / pi

  constructor(year?: number, month?: number, day?: number) {
    // Initialize constants
    this.YugaCivilDays = this.YugaRotation_star - this.YugaRotation_sun;
    
    if (year !== undefined && month !== undefined && day !== undefined) {
      this.year = year;
      this.month = month;
      this.day = day;
      const { gYear, gMonth, gDay } = this.toGregorian(year, month, day);
      this.englishDate = new Date(gYear, gMonth - 1, gDay);
    } else {
      // Default to current date
      this.englishDate = new Date();
      const { bsYear, bsMonth, bsDay } = this.fromGregorian(
        this.englishDate.getFullYear(),
        this.englishDate.getMonth() + 1,
        this.englishDate.getDate()
      );
      this.year = bsYear;
      this.month = bsMonth;
      this.day = bsDay;
    }
  }

  // Create a new BikramDate from a JavaScript Date
  static fromDate(date: Date): BikramDate {
    const bikramDate = new BikramDate();
    const { bsYear, bsMonth, bsDay } = bikramDate.fromGregorian(
      date.getFullYear(),
      date.getMonth() + 1,
      date.getDate()
    );
    bikramDate.year = bsYear;
    bikramDate.month = bsMonth;
    bikramDate.day = bsDay;
    return bikramDate;
  }

  // Convert to JavaScript Date
  toJsDate(): Date {
    return this.englishDate;
  }

  // Get year in Bikram Sambat
  getYear(): number {
    return this.year;
  }

  // Get month in Bikram Sambat (1-12)
  getMonth(): number {
    return this.month;
  }

  // Get day in Bikram Sambat
  getDate(): number {
    return this.day;
  }

  // Get days in a specific month
  getMonthDays(): number {
    if (this.year >= BS_START_YEAR && this.year <= BS_END_YEAR) {
      const yearIdx = this.year - BS_START_YEAR;
      if (yearIdx >= 0 && yearIdx < NP_MONTHS_DATA.length) {
        return NP_MONTHS_DATA[yearIdx][this.month - 1];
      }
    }
    return this.daysInMonth(this.year, this.month);
  }

  // Get the day of week (0-6, 0 = Sunday)
  getDay(): number {
    return this.englishDate.getDay();
  }

  // Convert AD (Gregorian) date to BS (Bikram Sambat)
  fromGregorian(gYear: number, gMonth: number, gDay: number): { bsYear: number; bsMonth: number; bsDay: number } {
    // Reference date: 1 Baisakh 2000 BS = 14 April 1943 AD
    const refDate = new Date(1943, 3, 14); // Month is 0-indexed in JS Date
    const targetDate = new Date(gYear, gMonth - 1, gDay);
    
    // Calculate difference in days
    const timeDiff = targetDate.getTime() - refDate.getTime();
    let daysDiff = Math.floor(timeDiff / (1000 * 3600 * 24));
    
    let bsYear = BS_START_YEAR;
    let bsMonth = 1;
    let bsDay = 1;
    
    // Find the Bikram date
    if (daysDiff >= 0) {
      while (bsYear < BS_END_YEAR) {
        const yearIdx = bsYear - BS_START_YEAR;
        if (yearIdx >= NP_MONTHS_DATA.length) break;
        
        for (let i = 0; i < 12; i++) {
          if (daysDiff < NP_MONTHS_DATA[yearIdx][i]) {
            bsMonth = i + 1;
            bsDay = daysDiff + 1;
            return { bsYear, bsMonth, bsDay };
          } else {
            daysDiff -= NP_MONTHS_DATA[yearIdx][i];
          }
        }
        bsYear++;
      }
      
      // If we're outside our data range, use the fallback astronomical calculation
      return this.fromGregorianAstronomical(gYear, gMonth, gDay);
    } else {
      // For dates before the reference date, we'll use the astronomical calculation
      return this.fromGregorianAstronomical(gYear, gMonth, gDay);
    }
  }

  // Convert BS (Bikram Sambat) date to AD (Gregorian)
  toGregorian(bsYear: number, bsMonth: number, bsDay: number): { gYear: number; gMonth: number; gDay: number } {
    // Reference date: 1 Baisakh 2000 BS = 14 April 1943 AD
    const refDate = new Date(1943, 3, 14); // Month is 0-indexed in JS Date
    
    // Check if the date is within our data range
    if (bsYear >= BS_START_YEAR && bsYear <= BS_END_YEAR) {
      let totalDays = 0;
      
      // Add days from years
      for (let year = BS_START_YEAR; year < bsYear; year++) {
        const yearIdx = year - BS_START_YEAR;
        if (yearIdx >= 0 && yearIdx < NP_MONTHS_DATA.length) {
          totalDays += NP_MONTHS_DATA[yearIdx][12]; // Total days in that year (at index 12)
        }
      }
      
      // Add days from months
      const yearIdx = bsYear - BS_START_YEAR;
      if (yearIdx >= 0 && yearIdx < NP_MONTHS_DATA.length) {
        for (let month = 0; month < bsMonth - 1; month++) {
          totalDays += NP_MONTHS_DATA[yearIdx][month];
        }
      }
      
      // Add days from current month
      totalDays += (bsDay - 1);
      
      // Calculate Gregorian date
      const targetDate = new Date(refDate.getTime() + totalDays * 24 * 60 * 60 * 1000);
      
      return {
        gYear: targetDate.getFullYear(),
        gMonth: targetDate.getMonth() + 1, // Convert from 0-indexed to 1-indexed
        gDay: targetDate.getDate()
      };
    } else {
      // Use fallback method for dates outside our range
      return this.toGregorianAstronomical(bsYear, bsMonth, bsDay);
    }
  }

  // Helper function to get days in month (matching C++ implementation)
  daysInMonth(bsYear: number, bsMonth: number): number {
    if (bsYear >= BS_START_YEAR && bsYear < BS_END_YEAR) {
      return NP_MONTHS_DATA[bsYear - BS_START_YEAR][bsMonth - 1];
    }
    // Fallback method
    const nextMonth = (bsMonth % 12) + 1;
    const nextYear = (bsMonth === 12) ? bsYear + 1 : bsYear;
    
    const { gYear: gYear1, gMonth: gMonth1, gDay: gDay1 } = this.toGregorian(bsYear, bsMonth, 1);
    const julian_date_start = this.getJulianDate(gYear1, gMonth1, gDay1);
    
    const { gYear: gYear2, gMonth: gMonth2, gDay: gDay2 } = this.toGregorian(nextYear, nextMonth, 1);
    const julian_date_end = this.getJulianDate(gYear2, gMonth2, gDay2);
    
    return Math.floor(julian_date_end - julian_date_start);
  }
  
  // Helper function to calculate Julian date (matching C++ implementation)
  private getJulianDate(year: number, month: number, day: number): number {
    if (month <= 2) {
      year -= 1;
      month += 12;
    }
    const a = Math.floor(year / 100.0);
    const b = 2 - a + Math.floor(a / 4.0);
    return Math.floor(365.25 * (year + 4716)) + Math.floor(30.6001 * (month + 1)) + day + b - 1524.5;
  }

  // Convert Julian date to Gregorian date
  private fromJulianDate(julian_date: number): { year: number; month: number; day: number } {
    const a = Math.floor(julian_date + 0.5);
    const b = a + 1537;
    const c = Math.floor((b - 122.1) / 365.25);
    const d = Math.floor(365.25 * c);
    const e = Math.floor((b - d) / 30.6001);
    const f = b - d - Math.floor(30.6001 * e) + (julian_date + 0.5 - a);
    
    const day = Math.floor(f);
    const month = (e < 14) ? (e - 1) : (e - 13);
    const year = (month > 2) ? (c - 4716) : (c - 4715);
    
    return { year, month, day };
  }

  // Astronomical calculation methods from the C++ implementation
  private getSauraMasaDay(ahar: number): { m: number; d: number } {
    if (this.todaySauraMasaFirstP(ahar)) {
      const day = 1;
      const tslong_tomorrow = this.getTslong(ahar + 1);
      let month = Math.floor(tslong_tomorrow / 30) % 12;
      month = (month + 12) % 12;
      return { m: month, d: day };
    } else {
      const yesterday = this.getSauraMasaDay(ahar - 1);
      return { m: yesterday.m, d: yesterday.d + 1 };
    }
  }

  private todaySauraMasaFirstP(ahar: number): boolean {
    let tslong_today = this.getTslong(ahar);
    let tslong_tomorrow = this.getTslong(ahar + 1);
    tslong_today -= Math.floor(tslong_today / 30) * 30;
    tslong_tomorrow -= Math.floor(tslong_tomorrow / 30) * 30;
    return (25 < tslong_today && tslong_tomorrow < 5) ? true : false;
  }

  private getTslong(ahar: number): number {
    let t1 = (this.YugaRotation_sun * ahar / this.YugaCivilDays);
    t1 -= Math.floor(t1);
    const mslong = 360 * t1;
    const x1 = mslong - this.PlanetApogee_sun;
    const y1 = this.PlanetCircumm_sun / 360;
    const y2 = Math.sin(x1 / this.rad);
    const y3 = y1 * y2;
    const x2 = Math.asin(y3) * this.rad;
    const x3 = mslong - x2;
    return x3;
  }

  // Fallback astronomical calculation methods
  private fromGregorianAstronomical(y: number, m: number, d: number): { bsYear: number; bsMonth: number; bsDay: number } {
    const julian = this.getJulianDate(y, m, d);
    const ahar = julian - 588465.5;
    
    const sauraMasaResult = this.getSauraMasaDay(ahar);
    const saura_masa_num = sauraMasaResult.m;
    const saura_masa_day = sauraMasaResult.d;
    
    const YearKali = Math.floor(ahar * this.YugaRotation_sun / this.YugaCivilDays);
    const YearSaka = YearKali - 3179;
    const nepalimonth = (saura_masa_num) % 12;
    const year = YearSaka + 135 + Math.floor((saura_masa_num - nepalimonth) / 12);
    const month = (saura_masa_num + 12) % 12 + 1;
    
    return {
      bsYear: year,
      bsMonth: month,
      bsDay: saura_masa_day
    };
  }

  private toGregorianAstronomical(bsYear: number, bsMonth: number, bsDay: number): { gYear: number; gMonth: number; gDay: number } {
    const YearSaka = bsYear - 135;
    const YearKali = YearSaka + 3179;
    let ahar = Math.floor((YearKali * this.YugaCivilDays) / this.YugaRotation_sun);
    
    // Adjust bsMonth to 0-11 for calculation
    const targetMonth = (bsMonth + 11) % 12;
    
    // Find the exact day
    let currentDay = this.getSauraMasaDay(ahar);
    
    // Try to find the exact date by moving forward or backward
    while (currentDay.m !== targetMonth || currentDay.d !== bsDay) {
      if (currentDay.m < targetMonth || (currentDay.m === targetMonth && currentDay.d < bsDay)) {
        ahar += 1;
      } else {
        ahar -= 1;
      }
      currentDay = this.getSauraMasaDay(ahar);
    }
    
    const julian_date = ahar + 588465.5;
    const gregorian = this.fromJulianDate(julian_date);
    
    return {
      gYear: gregorian.year,
      gMonth: gregorian.month, 
      gDay: gregorian.day
    };
  }
}
export { BS_START_YEAR, BS_END_YEAR };

File: src/bikram.cpp
/*
 * Copyright (C) khumnath
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "bikram.h"
#include "panchangacalculator.h"
#include "bsmonthdata.h"
#include <cmath>

#include "constants.h"

QString BikramCalendar::toDevanagari(const QString &numStr) {
  QString res;
  QString devNum = "०१२३४५६७८९";
  for (QChar c : numStr) {
    if (c.isDigit()) {
      res += devNum[c.digitValue()];
    } else {
      res += c;
    }
  }
  return res;
}

QString BikramCalendar::fromDevanagari(const QString &devanagariStr) {
  QString res;
  QString devNum = "०१२३४५६७८९";
  for (QChar c : devanagariStr) {
    int idx = devNum.indexOf(c);
    if (idx != -1) {
      res += QString::number(idx);
    } else {
      res += c;
    }
  }
  return res;
}

bool BikramCalendar::todaySauraMasaFirstP(const PanchangaCalculator* calc, double ahar) {
  double ahar_local_today = ahar - DESHANTARA;
  double ahar_local_tomorrow = ahar_local_today + 1.0;
  double tslong_today = calc->analyticalSunLong(ahar_local_today);
  double tslong_tomorrow = calc->analyticalSunLong(ahar_local_tomorrow);
  tslong_today -= std::floor(tslong_today / 30.0) * 30.0;
  tslong_tomorrow -= std::floor(tslong_tomorrow / 30.0) * 30.0;
  return (25.0 < tslong_today && tslong_tomorrow < 5.0);
}

void BikramCalendar::getSauraMasaDay(const PanchangaCalculator* calc, double ahar, int &month, int &day) {
  double currentAhar = std::floor(ahar);
  int daysElapsed = 0;
  while (daysElapsed < 40) {
    if (todaySauraMasaFirstP(calc, currentAhar)) {
      double tslong_tomorrow = calc->analyticalSunLong(currentAhar - DESHANTARA + 1.0);
      int m = static_cast<int>(std::floor(tslong_tomorrow / 30.0)) % 12;
      month = (m + 12) % 12;
      day = daysElapsed + 1;
      return;
    }
    currentAhar--;
    daysElapsed++;
  }
  month = 0;
  day = 1; // Fallback
}

int BikramCalendar::getBsYearFromAhar(const PanchangaCalculator* calc, double ahar) {
  int m, d;
  getSauraMasaDay(calc, ahar, m, d);
  double aharForYearCalc = ahar + (4 - m) * 30.0;
  int yearKali = static_cast<int>(std::floor(aharForYearCalc / SOLAR_YEAR_IN_DAYS));
  int yearSaka = yearKali - 3179;
  return yearSaka + 135;
}

int BikramCalendar::getBsYearFromAharAndMonth(double ahar, int m) {
  double aharForYearCalc = ahar + (4 - m) * 30.0;
  int yearKali = static_cast<int>(std::floor(aharForYearCalc / SOLAR_YEAR_IN_DAYS));
  int yearSaka = yearKali - 3179;
  return yearSaka + 135;
}

double BikramCalendar::findAharForBsDate(const PanchangaCalculator* calc, int bsYear, int monthIndex, int day) {
  int yearSaka = bsYear - 135;
  int yearKali = yearSaka + 3179;
  double ahar = std::floor(yearKali * SOLAR_YEAR_IN_DAYS);
  
  for (int i = 0; i < 5; i++) {
    int foundBsYear = getBsYearFromAhar(calc, ahar);
    int yearDifference = bsYear - foundBsYear;
    if (yearDifference == 0)
      break;
    ahar += std::round(yearDifference * SOLAR_YEAR_IN_DAYS);
  }
  
  int m, d;
  getSauraMasaDay(calc, ahar, m, d);
  double dayDifference = std::round((monthIndex * 30.5 + day) - (m * 30.5 + d));
  ahar += dayDifference;
  getSauraMasaDay(calc, ahar, m, d);
  
  int safetyCounter = 0;
  while (safetyCounter < 400) {
    int currentYear = getBsYearFromAharAndMonth(ahar, m);
    if (currentYear == bsYear && m == monthIndex && d == day) {
        break;
    }
    
    long currentTotal = currentYear * 12 + m;
    long targetTotal = bsYear * 12 + monthIndex;
    
    if (currentTotal < targetTotal || (currentTotal == targetTotal && d < day)) {
        ahar += 1.0;
    } else {
        ahar -= 1.0;
    }
    
    getSauraMasaDay(calc, ahar, m, d);
    safetyCounter++;
  }
  return ahar;
}

QDate BikramCalendar::fromBikramSambatAlgorithmic(const PanchangaCalculator* calc, int bsYearInput, int year, int monthIndex, int day) {
  double ahar = findAharForBsDate(calc, year, monthIndex, day);
  double jd = ahar + KaliEpoch;
  QDate adDate = fromJulianDay(jd);

  // Self-correction loop: attempt ±0..±4 day adjustments to find exact match
  int offsets[] = {0, 1, -1, 2, -2, 3, -3, 4, -4};
  for (int offset : offsets) {
    QDate testDate = adDate.addDays(offset);
    QVariantMap check = toBikramSambat(calc, testDate);
    int checkYear = check["year"].toInt();
    // Map the check year back to UI representation for comparison
    if (checkYear == bsYearInput && check["monthIndex"].toInt() == monthIndex && check["day"].toInt() == day) {
      return testDate;
    }
  }
  return adDate;
}

QVariantMap BikramCalendar::toBikramSambat(const PanchangaCalculator* calc, const QDate &date) {
  QVariantMap res;

  int daysFromEpoch = QDate(1943, 4, 14).daysTo(date);
  if (daysFromEpoch >= 0) {
    int tempYear = BsMonthData::BS_START_YEAR;
    int tempDays = daysFromEpoch;

    while (tempYear < BsMonthData::BS_START_YEAR + BsMonthData::NUM_YEARS) {
      int yearDays = BsMonthData::NP_MONTHS_DATA[tempYear - BsMonthData::BS_START_YEAR][12];
      if (tempDays < yearDays) {
        int tempMonth = 0;
        while (tempMonth < 12) {
          int monthDays = BsMonthData::NP_MONTHS_DATA[tempYear - BsMonthData::BS_START_YEAR][tempMonth];
          if (tempDays < monthDays) {
            res["year"] = tempYear;
            res["monthIndex"] = tempMonth;
            res["day"] = tempDays + 1;
            res["isComputed"] = false;
            res["isUnverified"] = (tempYear > 2083 && tempYear <= 2089);
            return res;
          }
          tempDays -= monthDays;
          tempMonth++;
        }
      }
      tempDays -= yearDays;
      tempYear++;
    }
  }

  // Fallback logic: custom toJulianDay for consistency
  double jd = toJulianDay(date);
  double ahar = jd - KaliEpoch; // jd is already at midnight (xxx.5)

  int m, d;
  getSauraMasaDay(calc, ahar, m, d);
  double aharForYearCalc = ahar + (4 - m) * 30.0;
  int yearKali = static_cast<int>(std::floor(aharForYearCalc / SOLAR_YEAR_IN_DAYS));
  int yearSaka = yearKali - 3179;
  int finalBsYear = yearSaka + 135;
  if (finalBsYear <= 0) finalBsYear -= 1; // Eliminate Year 0 for historical continuity

  res["year"] = finalBsYear;
  res["monthIndex"] = m;
  res["day"] = d;
  res["isComputed"] = true;
  res["isUnverified"] = false;
  return res;
}

QDate BikramCalendar::fromBikramSambat(const PanchangaCalculator* calc, int bsYearInput, int monthIndex, int day) {
  int year = bsYearInput < 0 ? bsYearInput + 1 : bsYearInput; // Map UI Year -1 back to internal Year 0
  int dataYear = year - BsMonthData::BS_START_YEAR;
  if (dataYear < 0 || dataYear >= BsMonthData::NUM_YEARS) {
    return fromBikramSambatAlgorithmic(calc, bsYearInput, year, monthIndex, day);
  }
  int totalDays = 0;
  for (int y = 0; y < dataYear; y++) {
    totalDays += BsMonthData::NP_MONTHS_DATA[y][12];
  }
  for (int m = 0; m < monthIndex; m++) {
    totalDays += BsMonthData::NP_MONTHS_DATA[dataYear][m];
  }
  totalDays += (day - 1);
  return QDate(1943, 4, 14).addDays(totalDays);
}

QVariantMap BikramCalendar::getBikramMonthInfo(const PanchangaCalculator* calc, int year, int monthIndex) {
  QVariantMap info;
  int dataYear = year - BsMonthData::BS_START_YEAR;

  if (dataYear >= 0 && dataYear < BsMonthData::NUM_YEARS) {
    int days = BsMonthData::NP_MONTHS_DATA[dataYear][monthIndex];
    int totalDays = 0;
    for (int y = 0; y < dataYear; y++) {
      totalDays += BsMonthData::NP_MONTHS_DATA[y][12];
    }
    for (int m = 0; m < monthIndex; m++) {
      totalDays += BsMonthData::NP_MONTHS_DATA[dataYear][m];
    }
    QDate startDate = QDate(1943, 4, 14).addDays(totalDays);

    info["year"] = year;
    info["monthIndex"] = monthIndex;
    info["monthName"] = solarMonthsList[monthIndex];
    info["totalDays"] = days;
    info["startAdDate"] = startDate;
    info["startDayOfWeek"] = getHistoricalWeekday(startDate);
    info["isComputed"] = false;
    return info;
  }

  // Astronomical fallback
  QDate firstDayAd = fromBikramSambat(calc, year, monthIndex, 1);
  int nextMonthIndex = (monthIndex + 1) % 12;
  int nextYear = (monthIndex == 11) ? year + 1 : year;
  QDate firstDayOfNextMonthAd = fromBikramSambat(calc, nextYear, nextMonthIndex, 1);

  int totalDays = firstDayAd.daysTo(firstDayOfNextMonthAd);

  info["year"] = year;
  info["monthIndex"] = monthIndex;
  info["monthName"] = solarMonthsList[monthIndex];
  info["totalDays"] = totalDays;
  info["startAdDate"] = firstDayAd;
  info["startDayOfWeek"] = getHistoricalWeekday(firstDayAd);
  info["isComputed"] = true;
  return info;
}

// Custom Julian Day conversions.

double BikramCalendar::toJulianDay(int year, int month0, int day) {
  // month0 is 0-indexed (0=January, 11=December), JS convention.
  int month1 = month0 + 1;
  int a = (14 - month1) / 12;
  // Use long long for intermediate calculations to prevent overflow for negative years
  long long y = (long long)year + 4800 - a;
  int m = month1 + 12 * a - 3;
  long long jdn = day + (153 * m + 2) / 5 + 365 * y + y / 4;

  // Hybrid: Gregorian after Oct 15, 1582; Julian before
  if (year > 1582 || (year == 1582 && month0 > 9) || (year == 1582 && month0 == 9 && day >= 15)) {
    jdn = jdn - y / 100 + y / 400 - 32045;
  } else {
    jdn = jdn - 32083;
  }
  return (double)jdn - 0.5; // midnight
}

double BikramCalendar::toJulianDay(const QDate &date) {
  return toJulianDay(date.year(), date.month() - 1, date.day());
}

QDate BikramCalendar::fromJulianDay(double jd) {
  // Proleptic Gregorian calendar for all dates
  jd += 0.5;
  double z = std::floor(jd);
  double alpha = std::floor((z - 1867216.25) / 36524.25);
  double a = z + 1 + alpha - std::floor(alpha / 4);
  double b = a + 1524;
  double c = std::floor((b - 122.1) / 365.25);
  double d = std::floor(365.25 * c);
  double e = std::floor((b - d) / 30.6001);
  int day = static_cast<int>(b - d - std::floor(30.6001 * e));
  int month = (e < 14) ? static_cast<int>(e) - 1 : static_cast<int>(e) - 13;
  int year = (month > 2) ? static_cast<int>(c) - 4716 : static_cast<int>(c) - 4715;
  return QDate(year, month, day);
}

int BikramCalendar::getHistoricalWeekday(const QDate &date) {
  double jd = toJulianDay(date);
  int z = static_cast<int>(std::floor(jd + 0.5));
  int wd = (z + 1) % 7;
  return wd >= 0 ? wd : wd + 7; // 0=Sunday..6=Saturday
}
